File: sdk/ruby/lib/authflow.rb
require 'net/http'
require 'json'
require 'uri'

module Authflow
  class Client
    attr_accessor :domain, :tenant_slug, :token

    def initialize(domain:, tenant_slug: nil)
      @domain = domain.chomp('/')
      @tenant_slug = tenant_slug
      @token = nil
    end

    def register(email:, password:, first_name:, last_name:, tenant_slug: nil)
      request(
        :post,
        '/auth/register',
        {
          email: email,
          password: password,
          firstName: first_name,
          lastName: last_name,
          tenantSlug: tenant_slug || @tenant_slug
        }
      )
    end

    def login(email:, password:, tenant_slug: nil)
      response = request(
        :post,
        '/auth/login',
        {
          email: email,
          password: password,
          tenantSlug: tenant_slug || @tenant_slug
        }
      )
      
      @token = response['token'] if response['token']
      response
    end

    def logout
      request(:post, '/auth/logout')
      @token = nil
      nil
    end

    def current_user
      response = request(:get, '/auth/me')
      response['user']
    end

    def enable_mfa_totp
      request(:post, '/user/mfa/totp/setup')
    end

    def verify_mfa_totp(code:)
      request(:post, '/user/mfa/totp/verify', { code: code })
    end

    def enable_mfa_sms
      request(:post, '/user/mfa/sms/enable')
    end

    def verify_mfa_sms(code:, remember_device: false)
      request(
        :post,
        '/user/mfa/sms/verify',
        {
          code: code,
          rememberDevice: remember_device
        }
      )
    end

    def send_magic_link(email:, redirect_url: nil, tenant_slug: nil)
      request(
        :post,
        '/auth/magic-link/send',
        {
          email: email,
          redirectUrl: redirect_url,
          tenantSlug: tenant_slug || @tenant_slug
        }
      )
    end

    def create_api_key(name:, expires_at: nil, permissions: nil)
      data = { name: name }
      data[:expiresAt] = expires_at if expires_at
      data[:permissions] = permissions if permissions
      
      request(:post, '/api-keys', data)
    end

    def list_api_keys
      request(:get, '/api-keys')
    end

    def delete_api_key(id:)
      request(:delete, "/api-keys/#{id}")
      nil
    end

    def create_webhook(url:, events:, description: nil)
      data = {
        url: url,
        events: events
      }
      data[:description] = description if description
      
      request(:post, '/webhooks', data)
    end

    def list_webhooks
      request(:get, '/webhooks')
    end

    def delete_webhook(id:)
      request(:delete, "/webhooks/#{id}")
      nil
    end

    private

    def request(method, endpoint, data = nil)
      uri = URI("#{@domain}/api#{endpoint}")
      
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == 'https'
      
      request_class = case method
                      when :get then Net::HTTP::Get
                      when :post then Net::HTTP::Post
                      when :put then Net::HTTP::Put
                      when :patch then Net::HTTP::Patch
                      when :delete then Net::HTTP::Delete
                      else raise ArgumentError, "Unsupported method: #{method}"
                      end
      
      request = request_class.new(uri.path)
      request['Content-Type'] = 'application/json'
      request['Authorization'] = "Bearer #{@token}" if @token
      request.body = data.to_json if data && method != :get
      
      response = http.request(request)
      
      raise "Request failed: #{response.code} #{response.body}" unless response.code.to_i < 400
      
      response.body.empty? ? {} : JSON.parse(response.body)
    end
  end
end

File: sdk/go/authflow.go
package authflow

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// Client represents an AuthFlow API client
type Client struct {
	Domain     string
	TenantSlug string
	HTTPClient *http.Client
	token      string
}

// User represents a user in the system
type User struct {
	ID            string    `json:"id"`
	Email         string    `json:"email"`
	FirstName     string    `json:"firstName"`
	LastName      string    `json:"lastName"`
	PhoneNumber   string    `json:"phoneNumber,omitempty"`
	Role          string    `json:"role"`
	TenantID      string    `json:"tenantId,omitempty"`
	EmailVerified bool      `json:"emailVerified"`
	PhoneVerified bool      `json:"phoneVerified"`
	MFAEnabled    bool      `json:"mfaEnabled"`
	CreatedAt     time.Time `json:"createdAt"`
}

// AuthResponse represents a successful authentication response
type AuthResponse struct {
	User  User   `json:"user"`
	Token string `json:"token"`
}

// RegisterData represents registration request data
type RegisterData struct {
	Email      string `json:"email"`
	Password   string `json:"password"`
	FirstName  string `json:"firstName"`
	LastName   string `json:"lastName"`
	TenantSlug string `json:"tenantSlug,omitempty"`
}

// LoginData represents login request data
type LoginData struct {
	Email      string `json:"email"`
	Password   string `json:"password"`
	TenantSlug string `json:"tenantSlug,omitempty"`
}

// NewClient creates a new AuthFlow client
func NewClient(domain string, tenantSlug string) *Client {
	return &Client{
		Domain:     domain,
		TenantSlug: tenantSlug,
		HTTPClient: &http.Client{Timeout: 30 * time.Second},
	}
}

// SetToken sets the authentication token
func (c *Client) SetToken(token string) {
	c.token = token
}

// request makes an HTTP request to the AuthFlow API
func (c *Client) request(method, endpoint string, body interface{}) (*http.Response, error) {
	url := fmt.Sprintf("%s/api%s", c.Domain, endpoint)

	var reqBody io.Reader
	if body != nil {
		jsonData, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		reqBody = bytes.NewBuffer(jsonData)
	}

	req, err := http.NewRequest(method, url, reqBody)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")
	if c.token != "" {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.token))
	}

	return c.HTTPClient.Do(req)
}

// Register creates a new user account
func (c *Client) Register(data RegisterData) (*User, error) {
	if data.TenantSlug == "" {
		data.TenantSlug = c.TenantSlug
	}

	resp, err := c.request("POST", "/auth/register", data)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("registration failed: %s", resp.Status)
	}

	var user User
	if err := json.NewDecoder(resp.Body).Decode(&user); err != nil {
		return nil, err
	}

	return &user, nil
}

// Login authenticates a user
func (c *Client) Login(data LoginData) (*AuthResponse, error) {
	if data.TenantSlug == "" {
		data.TenantSlug = c.TenantSlug
	}

	resp, err := c.request("POST", "/auth/login", data)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("login failed: %s", resp.Status)
	}

	var authResp AuthResponse
	if err := json.NewDecoder(resp.Body).Decode(&authResp); err != nil {
		return nil, err
	}

	c.token = authResp.Token
	return &authResp, nil
}

// Logout terminates the current session
func (c *Client) Logout() error {
	resp, err := c.request("POST", "/auth/logout", nil)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	c.token = ""
	return nil
}

// GetCurrentUser retrieves the current authenticated user
func (c *Client) GetCurrentUser() (*User, error) {
	resp, err := c.request("GET", "/auth/me", nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to get current user: %s", resp.Status)
	}

	var result struct {
		User User `json:"user"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return &result.User, nil
}

// EnableMFATOTP enables TOTP-based MFA
func (c *Client) EnableMFATOTP() (map[string]interface{}, error) {
	resp, err := c.request("POST", "/user/mfa/totp/setup", nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return result, nil
}

// VerifyMFATOTP verifies a TOTP code
func (c *Client) VerifyMFATOTP(code string) error {
	body := map[string]string{"code": code}
	resp, err := c.request("POST", "/user/mfa/totp/verify", body)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("MFA verification failed: %s", resp.Status)
	}

	return nil
}
